controller/controller.go

```diff
@@ -348,21 +356,64 @@ func (c controller) Destroy(ctx context.Context, req Request) (*task.Task, error
 }
 
 func (c controller) Update(ctx context.Context, req Request, opts UpdateOptions) (*task.Task, error) {
-	c.Config.Logger.Debug(ctx, "controller: handling update")
+	c.Config.Logger.Debug(ctx, "controller: handling update for group: %v", req.Group)
+
+	numRunning, err := c.getNumRunningSlices(ctx, req)
+	if err != nil {
+		return nil, maskAny(err)
+	}
+	c.Config.Logger.Debug(
+		ctx, "controller: running: %v, growth: %v, alive: %v, ready: %v",
+		numRunning, opts.MaxGrowth, opts.MinAlive, opts.ReadySecs,
+	)
+	updateAllowedRules := []struct {
+		// The human readable error message for this rule
+		message string
+		// broken is true when the update should not be allowed
+		broken bool
+	}{
+		{
+			message: "maximum units to create during update must be positive, or zero",
+			broken:  opts.MaxGrowth < 0,
+		},
+		{
+			message: "minimum alive units must be positive, or zero",
+			broken:  opts.MinAlive < 0,
+		},
+		{
+			message: "time between creating groups must be positive, or zero",
+			broken:  opts.ReadySecs < 0,
+		},
+		{
+			message: "cannot have minimum alive units greater than current number of units",
+			broken:  opts.MinAlive > numRunning,
+		},
+		{
+			message: "to keep all current units alive, max growth must be greater than 0",
+			broken:  opts.MinAlive == numRunning && opts.MaxGrowth < 1,
```

Comment:
I stayed with the "negative" validation, because this conditional felt awkward inverted. Would be welcome to feedback on it if we think the "positive" validation we talked about was much better.